src/libraries/Microsoft.PowerFx.Core/Functions/UserDefinedFunction.cs

```diff
@@ -0,0 +1,199 @@
+
+            public static INameResolver Merge(INameResolver globalNameResolver, RecordType parametersRecordType, INameResolver functionNameResolver)
+            {
+                return new UserDefinitionsNameResolver(globalNameResolver, ReadOnlySymbolTable.NewFromRecord(parametersRecordType), functionNameResolver);
+            }
+
+            private UserDefinitionsNameResolver(INameResolver globalNameResolver, INameResolver parameterNamedResolver, INameResolver functionNameResolver = null)
+            {
+                this._globalNameResolver = globalNameResolver;
+                this._parameterNamedResolver = parameterNamedResolver;
+                this._functionNameResolver = functionNameResolver;
+            }
+
+            public IExternalDocument Document => _globalNameResolver.Document;
+
+            public IExternalEntityScope EntityScope => _globalNameResolver.EntityScope;
+
+            public IExternalEntity CurrentEntity => _globalNameResolver.CurrentEntity;
+
+            public DName CurrentProperty => _globalNameResolver.CurrentProperty;
+
+            public DPath CurrentEntityPath => _globalNameResolver.CurrentEntityPath;
+
+            public TexlFunctionSet Functions => _functionNameResolver?.Functions ?? _globalNameResolver.Functions;
+
+            public bool SuggestUnqualifiedEnums => _globalNameResolver.SuggestUnqualifiedEnums;
+
+            public bool Lookup(DName name, out NameLookupInfo nameInfo, NameLookupPreferences preferences = NameLookupPreferences.None)
+            {
+                // lookup in the local scope i.e., function params & body and then look in global scope.
+                return _parameterNamedResolver.Lookup(name, out nameInfo, preferences) || _globalNameResolver.Lookup(name, out nameInfo, preferences);
+            }
+
+            public bool LookupDataControl(DName name, out NameLookupInfo lookupInfo, out DName dataControlName)
+            {
+                // params will not have any data controls, hence looking in just _globalNameResolver
+                return _globalNameResolver.LookupDataControl(name, out lookupInfo, out dataControlName);
+            }
+
+            public IEnumerable<TexlFunction> LookupFunctions(DPath theNamespace, string name, bool localeInvariant = false)
+            {
+                var functionsFromGlobalNameResolver = _globalNameResolver.LookupFunctions(theNamespace, name, localeInvariant);
+                if (_functionNameResolver == null)
+                {
+                    return functionsFromGlobalNameResolver;
+                }
+                else
+                {
+                    var functions = _functionNameResolver.LookupFunctions(theNamespace, name, localeInvariant);
+                    return functions.Any() ? functions : functionsFromGlobalNameResolver;
+                }
+            }
+
+            public IEnumerable<TexlFunction> LookupFunctionsInNamespace(DPath nameSpace)
```

Comment:
Nit: is this code identical to the code in the previous function? If so can they be shared?

Reply:
No they are not identical.